Add additonal type traits and functions

shrijitsingh99 opened the pull request.

shrijitsingh99 committed:
Rename executor_of_type type trait

shrijitsingh99 committed:
Add require type trait

kunaltyagi commented on application/include/mmul.hpp:
```diff
@@ -17,15 +17,14 @@
 
 template <typename Executor,
-          typename execution::executor_of_type<inline_executor, Executor> = 0>
+          typename execution::instance_of<inline_executor, Executor> = 0>
```
Rename to `B_is_instance_of_A` or something else which grammatically suggests that the second argument is instantiated template of arg 1

shrijitsingh99 replied:
How does `is_instance_of_base` sound? It's on the lines of `is_base_of`?

kunaltyagi commented on application/include/mmul.hpp:
```diff
@@ -42,18 +41,18 @@ void mmul(Executor ex, MatrixXd &a, MatrixXd &b, MatrixXd &c) {
 
-  std::array<int, 6> shape  {(int) ceil(a.rows() / 2.0), (int) ceil(b.cols() / 2.0), 1, 2, 2, 1};
+  std::array<int, 6> shape{
+      (int)ceil(a.rows() / 2.0), (int)ceil(b.cols() / 2.0), 1, 2, 2, 1};
```
Use C++ casts, elsewhere too

kunaltyagi commented on executor/include/execution/trait/can_require.hpp:
```diff
@@ -0,0 +1,32 @@
+
+template <typename Executor, typename Property,
+          typename std::enable_if_t<
+              Property::template is_applicable_property_v<Executor> &&
```
Bikeshedding: `is_property_applicable` instead?

shrijitsingh99 replied:
`is_applicable_property` is what the proposal has used, so using the same naming.

kunaltyagi commented on executor/include/execution/trait/can_require.hpp:
```diff
@@ -0,0 +1,32 @@
+}
+
+template <typename Executor, typename Properties, typename = void>
+struct can_require : std::false_type {};
```
What's the difference between `is_applicable_property` and `can_require`?

shrijitsingh99 replied:
`is_applicable_property` is used to restrict the mentioned properties to executors to only executors and not any type.

`can_require` checks whether the specified executor supports the specified require customization point.

shrijitsingh99 committed:
Make `is_executor` permanently true temporarily

Temporay change till the executor conditions mentioned (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1660r0.pdf) are used

shrijitsingh99 committed:
Add default bulk_execute functionality

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1660r0.pdf
Seems to indicate, a default bulk_execute will be provided which repeatedly calls execute

shrijitsingh99 committed:
Split and rename is_instance_of

shrijitsingh99 committed:
Fix formatting

kunaltyagi commented on executor/include/execution/executor/base_executor.hpp:
```diff
@@ -4,20 +4,17 @@
-    typename Interface, typename Cardinality, typename Blocking,
-    typename ProtoAllocator>
+template <template <typename...> typename Derived, typename Interface,
+          typename Cardinality, typename Blocking, typename ProtoAllocator>
```
Any reason for naming the parameters if you're not using them?

shrijitsingh99 replied:
I am using them in `query` member functions. There are some use cases for them which I think will come up.

kunaltyagi commented on executor/include/execution/executor/cuda_executor.hpp:
```diff
@@ -11,16 +11,18 @@
-template <typename Interface,typename Cardinality,typename Blocking, typename ProtoAllocator>
-struct cuda_executor: executor<sse_executor, Interface, Cardinality, Blocking, ProtoAllocator> {
-
+template <typename Interface, typename Cardinality, typename Blocking,
```
How about `<typename ...Args>` here too? It'll allow you to modify the executor class without worry of modifying the executors (just the usage, like in L44 and in consumer code)

shrijitsingh99 replied:
I need explicit access to the passed template arguments, like in the query functions in `executor` class. There will be some cases where I will need the argument name.

shrijitsingh99 committed:
Switch to C++ casts

shrijitsingh99 committed:
Add additional type traits

shrijitsingh99 committed:
Update executor design to match with proposal

shrijitsingh99 committed:
Switch to lambdas

kunaltyagi commented on application/include/mmul.hpp:
```diff
@@ -17,16 +17,17 @@
-  c = a * b;
+  auto mul = [](MatrixXd &a, MatrixXd &b, MatrixXd &c) { c = a * b; };
+
+  ex.execute(std::bind(mul, std::ref(a), std::ref(b), std::ref(c)));
```
`bind` ... over lambda? I do hope you plan on rectifying this 😄

shrijitsingh99 replied:
Fixed. Hadn't notice this.

kunaltyagi commented on application/include/mmul.hpp:
```diff
@@ -37,16 +38,21 @@ void mmul(Executor ex, MatrixXd &a, MatrixXd &b, MatrixXd &c) {
     }
   };
-  ex.bulk_execute(mul, 1, a, b, c);
+  ex.bulk_execute(std::bind(mul, std::ref(a), std::ref(b), std::ref(c),
```
Why not capture `a`, `b`, `c` in the lambda itself?

shrijitsingh99 replied:
Fixed

kunaltyagi reviewed (commented):
partial

shrijitsingh99 committed:
Remove bind in favour of lambda

shrijitsingh99 committed:
Fix is_executor trait bug

shrijitsingh99 committed:
Fix header

SergioRAgostinho commented on application/include/mmul.hpp:
```diff
@@ -17,17 +17,17 @@
 
 template <typename Executor,
-          typename execution::executor_of_type<inline_executor, Executor> = 0>
+          typename execution::instance_of_base<inline_executor, Executor> = 0>
```
Note to self: confirm later. 

This is what I would expect to see
```cpp
std:enable_if_t<execution::instance_of_base_v<inline_executor, Executor>, int> = 0
```
With the current syntax, you're shadowing what is happening and I see you assigning a 0 to something which I'm not sure what it is. 

~~proposal converge this closer to what is used in std rename to `is_child_of`, since you pass the derived type before the base type.~~

shrijitsingh99 replied:
I understand your point, the reason the name might be confusing is since I intentionally want to keep It short. This trait will be used everywhere so having a large name will pollute the codebase everywhere.

SergioRAgostinho replied:
`enable_as_int_if_instance_of_base_t` or `int_if_instance_of_base_t`? I think we need to compromise a little for clarity.

kunaltyagi replied:
Passing a 0 (or some meaningless value) to something is an alternative (and stronger) design for `typename X = default_type` to prevent the user from overloading the design. Once this is known, it's not that confusing. `enable_if_instance_of_template` (throwing my hat in the bikeshed) might be more expressive, but barely saves space (original intent). Another option might be `is_instance_of_template<...> = true`

> For new API, users demand expressiveness and bold syntax, but once they start using it, they want more concise syntax
- Bjarne (paraphrased)

Apparently, an age old problem

SergioRAgostinho commented on executor/include/execution/trait/is_instance_of_base.hpp:
```diff
@@ -0,0 +1,26 @@
+
+namespace execution {
+
+template <template <typename...> class Type, typename Executor>
+struct is_instance_of_base : std::false_type {};
+
+template <template <typename...> class Type, typename... Args>
+struct is_instance_of_base<Type, Type<Args...>> : std::true_type {};
```
What is the intention behind these methods? Should you not be using `std::is_base_of` here?

shrijitsingh99 replied:
The difference is that `is_instance_of_base` checks if it matches any of the instantiated template classes, i.e. both second arguments belong to the same template class mentioned in the first arg.
Eg:
`inline_executor<oneway_t, single_t, blocking_t::always_t>` and `inline_executor<twoway_t, single_t, blocking_t::never_t>` are both instances of the same class template `inline_executor<...>`. 
So the trait will return true for both the instances.

SergioRAgostinho replied:
The only thing you're effectively checking is if `Type` accepts at least one template parameters
https://godbolt.org/z/zVP7Qj

There's no notion of inheritance whatsoever being tested here. Using the name `instance_of_base` is misleading. Disregarding the actual implementation, can you describe in words what you're trying to check here?

kunaltyagi replied:
Use of `base` is deceptive here. `instance_of_template` might be more expressive

shrijitsingh99 replied:
Yeah, Kunal is right the name is misleading. The inheritance aspect is currently not there but I had some ideas regarding it. WIll bring it up again at a later stage.

SergioRAgostinho commented on executor/include/execution/executor/base_executor.hpp:
```diff
@@ -4,22 +4,33 @@
-template <template <typename, typename, typename, typename> typename Derived,
-    typename Interface, typename Cardinality, typename Blocking,
-    typename ProtoAllocator>
+template <template <typename...> typename Derived, typename Interface,
+          typename Cardinality, typename Blocking, typename ProtoAllocator>
```
Can't `Interface`, `Cardinality`, etc... be all inferred from `Derived`? i.e. inside the class you have something like
```cpp
using Derived::Interface;
using Derived::Cardinality;
```

Edit: I know we discussed this last time and your argument was
> Then every derived class needs to define these

I'm not convinced you need to pass the derived class back to base in this design, just for an equivalence operator.

shrijitsingh99 replied:
Created a new issue #13  to continue the discussion.

shrijitsingh99 commented:
If no additional issues will merge this. The remainder of the issues can be solved in #12

kunaltyagi commented on executor/include/execution/trait/is_executor.hpp:
```diff
@@ -4,24 +4,31 @@
-        void>> : std::true_type {};
+        std::is_copy_constructible_v<
+            std::remove_cv_t<std::remove_reference_t<Executor>>> &&
+            std::is_same_v<
```
Should this be out-dented?

kunaltyagi commented on executor/include/execution/trait/is_executor.hpp:
```diff
@@ -4,24 +4,31 @@
+            std::is_same_v<
+                decltype(std::declval<Executor>().execute(detail::noop)), void>,
+        decltype(std::declval<Executor&>() == std::declval<Executor&>(),
+                 (void)0)>> : std::true_type {};
```
Why is `void(0)` needed?

kunaltyagi replied:
Why is `(void)0` needed?

shrijitsingh99 replied:
To return the void type if the expression `std::declval<Executor&>() == std::declval<Executor&>()` is valid.
This article explains it better https://stackoverflow.com/questions/16044514/what-is-decltype-with-two-arguments

Just syntactic sugar in the end.

kunaltyagi replied:
Please rethink the way you've constructed the template. My concerns are:
* `is_executor` accepts a `Function` but doesn't test it. Is the function even needed? `is_executor_v` doesn't seem to need or expose the function
* The condition needs to be broken in 2 parts of `enable_if`
* Conditions are hard to understand. They should be refactored into named requirements so it's easier to read and understand

```cpp

template <class Candidate>
struct is_executor: std::conditional<
detail::all_objects_are_equal<Candidate> && detail::accepts_void_return_no_param_functor<Candidate>
,std::true_type, std::false_type> {};
```

My names are ofc bad (the papers you've read might have better names), but I hope this shows some direction which would be easier to understand

kunaltyagi commented on executor/include/execution/trait/can_require.hpp:
```diff
@@ -0,0 +1,33 @@
+template <typename Executor, typename Property>
+struct can_require<Executor, Property,
+                   std::void_t<decltype(execution::require(
+                       std::declval<Executor>(), std::declval<Property>()))>>
```
Maybe explain what you're checking: `require(executor, property)`

shrijitsingh99 replied:
Got it, will add some documentation, once the API is more stable.

kunaltyagi commented on executor/include/execution/trait/can_prefer.hpp:
```diff
@@ -0,0 +1,34 @@
+template <typename Executor, typename Property>
+struct can_prefer<Executor, Property,
+                  std::void_t<decltype(execution::require(
+                      std::declval<Executor>(), std::declval<Property>()))>>
```
Is `can_prefer` and `can_require` the same?

shrijitsingh99 replied:
No, it seems like a bug.
Fixed in 089ef0421296e37ee29b72e8ffb3383bd6ab0fdc

kunaltyagi commented on executor/include/execution/executor/base_executor.hpp:
```diff
@@ -4,22 +4,33 @@
+  template <typename Executor,
+            typename execution::instance_of_base<Derived, Executor> = 0>
+  bool operator!=(const Executor &) const noexcept {
+    return std::is_same_v<
```
o.O ?

SergioRAgostinho replied:
you should probably just negate your equivalence operator.

shrijitsingh99 replied:
Oops, Fixed 293f980bea87c88d2c7f862f8916d40dd8a32a69

SergioRAgostinho commented on application/include/mmul.hpp:
```diff
@@ -17,17 +17,17 @@
 template <typename Executor,
-          typename execution::executor_of_type<inline_executor, Executor> = 0>
+          typename execution::instance_of_base<inline_executor, Executor> = 0>
 void mmul(Executor ex, MatrixXd &a, MatrixXd &b, MatrixXd &c) {
```
Two of these should be parameters should be const.

SergioRAgostinho replied:
Sorry for asking this again, what were the argument for not using tag dispatching based overloads? This feels like a solid use case for it.

shrijitsingh99 replied:
Tag dispatching overload will only allow executors with a fixed set of properties and type to be passed.
Eg: 
```cpp
void mmul(inline_executor<oneway_t, single_t, blocking_t::always_t> &ex, MatrixXd &a, MatrixXd &b, MatrixXd &c){...}
```
The above function will not accept an inline_executor having a custom allocator.

Tag dispatching will certainly be useful in some cases, let me come back to this point on where it can be useful.

SergioRAgostinho replied:
If you want the flexibility, you can always template the method for the specific properties which you don't care about the type, in your example case you claimed you don't care about the allocator property of the executor

```cpp
template<typename allocator_t>
void mmul(inline_executor<oneway_t, single_t, blocking_t::always_t, allocator_t> &ex, MatrixXd &a, MatrixXd &b, MatrixXd &c){...}
```

SergioRAgostinho commented on application/include/mmul.hpp:
```diff
@@ -38,22 +38,32 @@ void mmul(Executor ex, MatrixXd &a, MatrixXd &b, MatrixXd &c) {
-    typename execution::executor_of_type<cuda_executor, Executor> = 0>
+          typename execution::instance_of_base<cuda_executor, Executor> = 0>
 void mmul(Executor ex, MatrixXd &a, MatrixXd &b, MatrixXd &c) {
+  double *a_g, *b_g, *c_g;
+  a_g = static_cast<double *>(
+      device_upload(static_cast<void *>(a.data()), a.size() * sizeof(double)));
+  b_g = static_cast<double *>(
+      device_upload(static_cast<void *>(b.data()), b.size() * sizeof(double)));
+  c_g = static_cast<double *>(
+      device_upload(static_cast<void *>(c.data()), c.size() * sizeof(double)));
```
I think you only need the static cast for the return type, and you can initialize your pointer directly.

shrijitsingh99 replied:
Fixed in #12

SergioRAgostinho commented on application/include/mmul.hpp:
```diff
@@ -38,22 +38,32 @@ void mmul(Executor ex, MatrixXd &a, MatrixXd &b, MatrixXd &c) {
-  memcpy((double *)c.data(), c_g, 9 * sizeof(double));
-}
\ No newline at end of file
+  memcpy(static_cast<double *>(c.data()), c_g, 9 * sizeof(double));
```
no need for the cast I believe.

shrijitsingh99 replied:
Should be resolved in #12

SergioRAgostinho commented on executor/include/execution/executor/cuda_executor.hpp:
```diff
@@ -9,39 +9,50 @@
+  template <typename F>
+  void execute(F &&f) {
+#ifdef CUDA
+    cudaLaunchKernel((void *)f, 1, 1, nullptr, 0, 0);
```
c++-style casts

shrijitsingh99 replied:
Fixed in #12

SergioRAgostinho reviewed (commented):
Partial (7/17)

SergioRAgostinho commented on executor/include/execution/executor/base_executor.hpp:
```diff
@@ -4,22 +4,33 @@
+          typename Cardinality, typename Blocking, typename ProtoAllocator>
 class executor {
  public:
+  template <typename Executor,
+            typename execution::instance_of_base<Derived, Executor> = 0>
+  bool operator==(const Executor &) const noexcept {
+    return std::is_same_v<
+        Derived<Interface, Cardinality, Blocking, ProtoAllocator>, Executor>;
+  }
```
Binary and symmetric operators (sum, mult, equivalence, etc...) are usually recommended to bedefined outside the class.

shrijitsingh99 replied:
Any reason for this?

SergioRAgostinho replied:
http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Ro-symmetric
https://www.modernescpp.com/index.php/c-core-guidelines-rules-for-overloading-and-overload-operators
In your particular case the most pressing reason is because it  removes the need of passing the derived class as a template parameter to your base type.

SergioRAgostinho commented on executor/include/execution/property/base_property.hpp:
```diff
@@ -4,25 +4,25 @@
   template <class Executor>
   static constexpr auto static_query()
-  -> decltype(Executor::query(std::declval<Derived>())) {
+      -> decltype(Executor::query(std::declval<Derived>())) {
```
All query methods will always return bool right? No need to complicate.

SergioRAgostinho replied:
Have you considered defining this method outside the class as a free-form function with two template parameters. renamed simply `query`?

shrijitsingh99 replied:
> All query methods will always return bool right?

Actually no, if you query the property `ProtoAllocator` it should return the allocator itself. I haven't implemented it currently but it is to be done.

> defining this method outside the class 

Like `execution::query()`? As that is already defined in `can_query.hpp`, currently, the code offers query both as a member function and free function.

SergioRAgostinho replied:
> > All query methods will always return bool right?
> 
> Actually no, if you query the property `ProtoAllocator` it should return the allocator itself. I haven't implemented it currently but it is to be done.

Isn't that inconsistent with the rest of your query functions? All of them return `bool` except one. What's the reason for this different behavior?

> > defining this method outside the class
> 
> Like `execution::query()`? As that is already defined in `can_query.hpp`, currently, the code offers query both as a member function and free function.

I'm ok with the first two
```cpp
query<Executor, Property>(); //ok
executor.query<property>(); //ok
Property::static_query<Executor>(); // This forces you to have CRTP
```
You're imposing CRTP into your class design just to offer that method, and that's the trade-off I don't agree

shrijitsingh99 replied:
Let me revaluate why the TS specifies that and come up with a  convincing argument for its need.

SergioRAgostinho reviewed (commented):
When you write these classes. Can you identify things which are required to exist by the technical proposals?

There's no point in defining everything as a `class` if then your going to immediately set the visibility to `public`. `struct`s gives you that for free.

SergioRAgostinho commented on executor/include/execution/trait/is_executor.hpp:
```diff
@@ -4,24 +4,31 @@
+            std::remove_cv_t<std::remove_reference_t<Executor>>> &&
+            std::is_same_v<
+                decltype(std::declval<Executor>().execute(detail::noop)), void>,
+        decltype(std::declval<Executor&>() == std::declval<Executor&>(),
+                 (void)0)>> : std::true_type {};
```
I'm on this party as well. Can you explain in detail what is happening in these two lines.

shrijitsingh99 replied:
https://github.com/shrijitsingh99/unified-executors/pull/7#discussion_r448564245

SergioRAgostinho commented on executor/include/execution/trait/is_executor.hpp:
```diff
@@ -4,24 +4,31 @@
+}
 
-template <class Executor>
+template <typename Executor, typename Function>
```
You've added the `Function` parameter, but there's no explicit use of it. That usually, means it's not required.

shrijitsingh99 replied:
You are right, it isn't required.
Removed in fe249605c8ba21fd1ea17b7132f5d246605778c6

SergioRAgostinho commented on executor/include/execution/trait/is_instance_of_base.hpp:
```diff
@@ -0,0 +1,26 @@
+    is_instance_of_base<Type, Executor>::value;
+
+template <template <typename...> class Type, typename Executor>
+using instance_of_base =
```
This needs to be rename to something more expressive e.g. `enable_as_int_if_instance_of_base`. Otherwise it looks like magic whenever it is used.

shrijitsingh99 replied:
The idea behind it was to be magically :wink: so that It just works since this is going to be used everywhere.
Giving a long name would really something would just make template errors longer and add additionally increase the length of code in every executor function.

SergioRAgostinho replied:
I don't agree with this trade-off. Please come up with a more expressive name. I'm ok with `enable_if_executor_t` or something along those lines, but not with the current proposal.

shrijitsingh99 replied:
Got it. Will come up with something which explains the functionality better.

SergioRAgostinho reviewed (commented):
Final one.

shrijitsingh99 commented:
> When you write these classes. Can you identify things which are required to exist by the technical proposals?

I am not sure I understood this. Could you please elaborate?

> There's no point in defining everything as a class if then your going to immediately set the visibility to public. structs gives you that for free.

Got it, I don't think I need classes anywhere as of now. WIll convert everything to structs.

SergioRAgostinho commented:
> > When you write these classes. Can you identify things which are required to exist by the technical proposals?
> 
> I am not sure I understood this. Could you please elaborate?

Example: This method inside `basic_executor_property`
```cpp
  template <class T>
  static constexpr bool is_applicable_property() {
    return execution::is_executor<T>();
  }
```
looks useless to me. However, I think later you replied to Kunal, that it was part of the proposal and therefore had to exist. Knowing in advance through a comment that certain aspects of the API are mandatory, changes my review context. I don't have time to review the technical proposals, so knowing  certain things are required to exist prevents me from applying the wrong judgement.

The less cumbersome way out of this is, to keep a summarized list of things that the TS defines as mandatory for each class you implemented, that I can refer to. We can populate this list every time one of its items is brought up during reviews.

kunaltyagi commented on executor/include/execution/trait/can_prefer.hpp:
```diff
@@ -0,0 +1,34 @@
+//
+// Created by Shrijit Singh on 17/06/20.
```
Does it make sense to split `can_require` and `can_prefer` implementations? Especially since the implementation for both is exactly the same.

shrijitsingh99 replied:
That was a mistake they have slightly different implementations. It has been fixed in #12

kunaltyagi reviewed (commented):
Lots of code left :sweat_smile:

shrijitsingh99 commented:
Closing this in favor of #12. This has become quite outdated.

shrijitsingh99 commented:
> The less cumbersome way out of this is, to keep a summarized list of things that the TS defines as mandatory for each class you implemented, that I can refer to

Got it. WIll add appropriate comments everywhere.

shrijitsingh99 commented:
> Conditions are hard to understand. They should be refactored into named requirements so it's easier to read and understand

Yeah, it's becoming quite convoluted will break it down
